crates/ra_hir/src/ty.rs

```diff
@@ -1042,12 +1046,24 @@ impl<'a, D: HirDatabase> InferenceContext<'a, D> {
             },
             Expr::Tuple { exprs } => {
                 let mut ty_vec = Vec::with_capacity(exprs.len());
-                for arg in exprs.iter() {
-                    ty_vec.push(self.infer_expr(*arg, &Expectation::none())?);
+                for expr in exprs.iter() {
+                    ty_vec.push(self.infer_expr(*expr, &Expectation::none())?);
                 }
 
                 Ty::Tuple(Arc::from(ty_vec))
             }
+            Expr::Array { exprs } => {
+                let mut ty = Ty::Unknown;
```

Comment:
I think there's a nicer way to do this: create a new type variable here (you could just write `self.insert_type_vars(Ty::Unknown)`), and then pass that as the expected type to each of the sub expressions.

As a bonus, you can actually use the expected type you currently have. So instead of starting with `Ty::Unknown`, you can look into `expected.ty` and see if it's a slice or array type. If it is, you can use the element type from that.